test/smallarrayshrink.jl

```diff
@@ -0,0 +1,45 @@
+@testset "shrink small array" begin
+    x = [1, 2, 3, 4]
+    @test x[1] == 1
+    @test x[2] == 2
+    @test x[3] == 3
+    @test x[4] == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 0) == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 1) == 4
+    sizehint!(x, 10000)
+    @test x[1] == 1
+    @test x[2] == 2
+    @test x[3] == 3
+    @test x[4] == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 0) == 4
+    @test ccall(:jl_array_size, Int, (Any, UInt), x, 1) == 10000
```

Comment:
This is quite "clever"/obfuscated... took me a while to figure out. This is relying on the fact that `maxsize` is stored after `nrows` and that this function doesn't check its arguments. We would need a different way to do this.